sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala

```diff
@@ -0,0 +1,351 @@
+    outstandingZoneIds.foreach { zid =>
+      withClue(s"zone id = $zid") {
+        withDefaultTimeZone(zid) {
+          val start = instantToMicros(LocalDateTime.of(1, 1, 1, 0, 0, 0)
+            .atZone(zid)
+            .toInstant)
+          val end = instantToMicros(LocalDateTime.of(2100, 1, 1, 0, 0, 0)
+            .atZone(zid)
+            .toInstant)
+          var micros = start
+          do {
+            val rebased = rebaseGregorianToJulianMicros(zid, micros)
+            val rebasedAndOptimized = rebaseGregorianToJulianMicros(micros)
+            assert(rebasedAndOptimized === rebased)
+            micros += (MICROS_PER_MONTH * (0.5 + Math.random())).toLong
+          } while (micros <= end)
+        }
+      }
+    }
+  }
+
+  test("optimization of micros rebasing - Julian to Gregorian") {
+    outstandingZoneIds.foreach { zid =>
+      withClue(s"zone id = $zid") {
+        withDefaultTimeZone(zid) {
+          val start = rebaseGregorianToJulianMicros(
+            instantToMicros(LocalDateTime.of(1, 1, 1, 0, 0, 0).atZone(zid).toInstant))
+          val end = rebaseGregorianToJulianMicros(
+            instantToMicros(LocalDateTime.of(2100, 1, 1, 0, 0, 0).atZone(zid).toInstant))
+          var micros = start
+          do {
+            val rebased = rebaseJulianToGregorianMicros(zid, micros)
+            val rebasedAndOptimized = rebaseJulianToGregorianMicros(micros)
+            assert(rebasedAndOptimized === rebased)
+            micros += (MICROS_PER_MONTH * (0.5 + Math.random())).toLong
+          } while (micros <= end)
+        }
+      }
+    }
+  }
+
+  private def generateRebaseJson(
+      adjustFunc: Long => Long,
+      rebaseFunc: (ZoneId, Long) => Long,
+      dir: String,
+      fileName: String): Unit = {
+    import java.nio.file.{Files, Paths}
+    import java.nio.file.StandardOpenOption
+
+    import scala.collection.mutable.ArrayBuffer
+
+    import com.fasterxml.jackson.databind.ObjectMapper
+    import com.fasterxml.jackson.module.scala.{DefaultScalaModule, ScalaObjectMapper}
+
+    case class RebaseRecord(tz: String, switches: Array[Long], diffs: Array[Long])
+
+    val result = new ArrayBuffer[RebaseRecord]()
+    ALL_TIMEZONES.foreach { zid =>
+      withDefaultTimeZone(zid) {
+        val start = adjustFunc(instantToMicros(LocalDateTime.of(1, 1, 1, 0, 0, 0)
+          .atZone(zid)
+          .toInstant))
+        val end = adjustFunc(instantToMicros(LocalDateTime.of(2100, 1, 1, 0, 0, 0)
+          .atZone(zid)
+          .toInstant))
+
+        var micros = start
+        var diff = Long.MaxValue
+        var counter = 0
+        val maxStep = MICROS_PER_MONTH
+        var step: Long = MICROS_PER_SECOND
+        val switches = new ArrayBuffer[Long]()
+        val diffs = new ArrayBuffer[Long]()
+        while (micros < end) {
+          val rebased = rebaseFunc(zid, micros)
+          val curDiff = rebased - micros
+          if (curDiff != diff) {
```

Comment:
so we always hit this branch for the first record? because the `diff` is `Long.MaxValue` at the beginning.

Reply:
yes, we need the start point and diff for it.